cyipopt/cython/ipopt_wrapper.pyx

```diff
@@ -654,6 +654,102 @@ cdef class Problem:
 
         return np_x, info
 
+    def get_current_iterate(self, scaled=False):
+        # Allocate arrays to hold the current iterate
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_x
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_x_L
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_x_U
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_g
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_g
+        np_x = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_mult_x_L = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_mult_x_U = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_g = np.zeros((self.__m,), dtype=DTYPEd).flatten()
+        np_mult_g = np.zeros((self.__m,), dtype=DTYPEd).flatten()
+
+        # Cast to C data types
+        x = <Number*>np_x.data
+        mult_x_L = <Number*>np_mult_x_L.data
+        mult_x_U = <Number*>np_mult_x_U.data
+        g = <Number*>np_g.data
+        mult_g = <Number*>np_mult_g.data
+
+        # NOTE: GetIpoptCurrentIterate can *only* be called during an
+        # intermediate callback (otherwise __nlp->tnlp->ip_data_ is NULL)
+        # TODO: Either catch error or avoid calling if we are not in an
+        # intermediate callback
+        ret = GetIpoptCurrentIterate(
+            self.__nlp,
+            scaled,
+            self.__n,
+            x,
+            mult_x_L,
+            mult_x_U,
+            self.__m,
+            g,
+            mult_g,
+        )
+
+        # Return values to user
+        # - Is another data type (e.g. dict, namedtuple) more appropriate than
+        #   simply a tuple?
```

Comment:
In `.solve()` we return a tuple of an ndarray (optimal solution) and a dict mapping to other arrays including the optimal solution again. I can't say that is the cleanest API. But maybe we should mimic it here, since it is similar information?

Reply:
We could mimic this for `get_current_iterate`, but I don't see what the first ndarray should be for `get_current_violations`, and I think it's more important for these to be consistent with each other than with solve. I'm currently returning dicts for both methods. In `get_current_iterate`, the keys are a subset of those in solve's returned dict. In `get_current_violations`, they match the GetIpoptCurrentIterate [documentation](https://coin-or.github.io/Ipopt/IpStdCInterface_8h.html#acf68350650b7a893061e110ab0e9264f), although I'm thinking about changing `"nlp_constraint_violation" -> "g_violation"`.